src/audio/castleinternalsoundfile.pas

```diff
@@ -90,7 +90,58 @@   TSoundFile = class
     procedure ConvertTo16bit; virtual;
   end;
 
+  { TStreamedSoundFile }
```

Comment:
I was thinking about this and it's hard for me to create some reasonable base class, the use of which will not require continuous casting to the sub-type (as in the case of `TAbstractSoundFile.CreateFromFile(...):TAbstractSoundFile`). I think the differences in the method of use are too big for me to connect these two branches. A common ancestor will always require casting after createFromFile() or class responsibility will be unclean (by adding all methods and implement only some of them). Unless I do not see any obvious solution, that you see.

BTW. I know it's quite annoying to have two classes for each sound format like `TSoundOggVorbis` and `TStreamedSoundOggVorbis`. But I think it's easier to maintain two simple classes (with well-defined responsibility) than one with a lot of "if streming then" conditions.

Reply:
I will need to think about it more :)

I understand the case you make. Indeed we want to have a clean code -- and I can see that sometimes (like in this case) it may be cleaner (producing code simpler to understand) to duplicate some stuff, than to try to reuse everything by a common class.

So, I think you're right here, and I'll try to do some thinking about it later :), maybe I'll be able to propose something.

Reply:
Done in https://github.com/castle-engine/castle-engine/pull/127/commits/80c2989c7853a83cac9d36f54ab8a5588f666f8e